apps/btc_app/get_public_key.c

```diff
@@ -0,0 +1,292 @@
+ * @param public_key Storage location for raw uncompressed public key
+ * @param address Storage location for encoded public address
+ *
+ * @return int length of the derived public address
+ * @retval 0 If derivation failed
+ */
+static int btc_get_address(const uint8_t *seed,
+                           const uint32_t *path,
+                           uint32_t path_length,
+                           uint8_t *public_key,
+                           char *address);
+
+/**
+ * @brief Takes and sends the uncompressed public key to the host as
+ * btc_result_t
+ *
+ * @param public_key An immutable reference to uncompressed public key buffer
+ */
+static void send_public_key(const uint8_t *public_key);
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+static bool check_which_request(const btc_query_t *query,
+                                pb_size_t which_request) {
+  if (which_request != query->get_public_key.which_request) {
+    btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_REQUEST);
+    return false;
+  }
+
+  return true;
+}
+
+static bool validate_request_data(btc_get_public_key_request_t *request) {
+  bool status = true;
+
+  // TODO: Enable btc/coin specific check
+  if (false ==
+      verify_receive_derivation_path(request->initiate.derivation_path,
+                                     request->initiate.derivation_path_count)) {
+    btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_DATA);
+    status = false;
+  }
+  return status;
+}
+
+static int btc_get_address(const uint8_t *seed,
+                           const uint32_t *path,
+                           uint32_t path_length,
+                           uint8_t *public_key,
+                           char *address) {
+  HDNode node = {0};
+  uint8_t address_version = 0;
+  uint32_t purpose_index = path[0];
+  uint32_t coin_index = path[1];
+  char addr[50] = "";
+  int address_length = 0;
+
+  if (!derive_hdnode_from_path(
+          path, path_length, SECP256K1_NAME, seed, &node)) {
+    // send unknown error; unknown failure reason
+    btc_send_error(ERROR_COMMON_ERROR_UNKNOWN_ERROR_TAG, 1);
+    memzero(&node, sizeof(HDNode));
+    return address_length;
+  }
+
+  switch (purpose_index) {
+    case NATIVE_SEGWIT:
+      // ignoring the return status and handling by size of address
+      get_segwit_address(
+          node.public_key, sizeof(node.public_key), coin_index, addr);
+      break;
+    case NON_SEGWIT:
+      get_version(purpose_index, coin_index, &address_version, NULL);
+      hdnode_get_address(&node, address_version, addr, 35);
+      break;
+    // TODO: add support for taproot and segwit
+    default:
+      break;
+  }
+
+  address_length = strnlen(addr, sizeof(addr));
+  if (0 >= address_length) {
```

Comment:
Can `address_length` ever be < 0?

I think strnlen returns `size_t` which ultimately should be of `uint_x` type

Reply:
thanks. missed this